File: sample.java
import java.io.File;
import java.io.IOException;
import com.mpatric.mp3agic.Mp3File;
import com.mpatric.mp3agic.ID3v1;
import com.mpatric.mp3agic.ID3v1Tag;
import com.mpatric.mp3agic.ID3v2;
import com.mpatric.mp3agic.ID3v24Tag;
import com.mpatric.mp3agic.UnsupportedTagException;
import com.mpatric.mp3agic.NotSupportedException;
import com.mpatric.mp3agic.InvalidDataException;
import java.util.List;
import java.util.ArrayList;

public class sample {
	public static void main(String[] args) {
		String musicDir = System.getenv("HOME");
		musicDir = musicDir + "/Music";
		String[] artistsBuffer;
		String[] albumsBuffer;
		String[] musicBuffer;

		try {
			
			// gets artists
			artistsBuffer = getDir(new File(musicDir));
			
			// iterate artists
			for (String artistsIterator : artistsBuffer) {
				musicBuffer = getMusic(new File(musicDir + "/" + artistsIterator));
				System.out.println("artist:\t" + artistsIterator);
				// edit mp3 files from artists
				for (String musicIterator : musicBuffer) {
					editMusic(musicDir, artistsIterator, "", musicIterator);
					System.out.println("title:\t" + musicIterator.substring(0, musicIterator.length() - 4));
				}

				// gets albums from artists
				albumsBuffer = getDir(new File(musicDir + "/" + artistsIterator));
				// edits music from an album from an arist
				if (albumsBuffer != null) {
					for (String albumsIterator : albumsBuffer) {
						musicBuffer = getMusic(new File(musicDir + "/" + artistsIterator + "/" + albumsIterator));
						System.out.println("album:\t" + albumsIterator);
						if (musicBuffer != null) {
							for (String musicIterator : musicBuffer) {
								editMusic(musicDir, artistsIterator, albumsIterator, musicIterator);
								System.out.println("title:\t" + musicIterator.substring(0, musicIterator.length() - 4));
							}
						}
					}
				}
				System.out.println();
			}
		} catch (IOException e) {
			e.printStackTrace();
		} catch (UnsupportedTagException e) {
			e.printStackTrace();
		} catch (NotSupportedException e) {
			e.printStackTrace();
		} catch (InvalidDataException e) {
			e.printStackTrace();
		}
	}

	public static String[] getDir(File dir) { // gets directories
		List<String> albumList = new ArrayList<String>();
		String[] buffer;
		if (dir.listFiles() != null) {
			for (File albums : dir.listFiles()) {
				if (albums.isDirectory()) {
					albumList.add(albums.getName());
				}
			}
		}
		buffer = new String[albumList.size()];
		albumList.toArray(buffer);
		return buffer;
	}

	public static String[] getMusic(File dir) { // gets .mp3 files
		List<String> musicList = new ArrayList<String>();
		String[] buffer;
		if (dir.listFiles() != null) {
			for (File music : dir.listFiles()) {
				if (music.isFile()) {
					if (music.getName().substring(music.getName().length() - 3, music.getName().length()).equals("mp3")) {
						musicList.add(music.getName());
					}
				}
			}
		}
		buffer = new String[musicList.size()];
		musicList.toArray(buffer);
		return buffer;
	}

	public static void editMusic(String dir, String artist, String album, String file) throws IOException, UnsupportedTagException, NotSupportedException, InvalidDataException {
		Mp3File mp3File;
		if (!file.substring(file.length() - 3, file.length()).equals("mp3")) {
				return;
		}
		/* System.out.println("artist: " + artist);
		System.out.println("album: " + album);
		System.out.println("title: " + file); */
		if (album.equals("")) {
			mp3File = new Mp3File(dir + "/" + artist + "/" + file);
		} else {
			mp3File = new Mp3File(dir + "/" + artist + "/" + album + "/" + file);
		}
		if (file.substring(file.length() - 4, file.length()).equals(".mp3")) {
			file = file.substring(0, file.length() - 4);
		}
		
		if (mp3File.hasId3v1Tag()) {
			ID3v1 id3v1Tag = new ID3v1Tag();
			id3v1Tag.setArtist(artist);
			id3v1Tag.setAlbum(album);
			id3v1Tag.setTitle(file);
			mp3File.setId3v1Tag(id3v1Tag);
		} else if (mp3File.hasId3v2Tag()) {
			ID3v2 id3v2Tag = new ID3v24Tag();
			id3v2Tag.setArtist(artist);
			id3v2Tag.setAlbum(album);
			id3v2Tag.setTitle(file);
			mp3File.setId3v2Tag(id3v2Tag);
		}
		/* if (mp3File.hasCustomTag()) {
			System.out.println("id3v1");
		  mp3File.removeCustomTag();
			} */
		if (album.equals("")) {
			mp3File.save(dir + "/" + artist + "/" + file + ".mp3");
		} else {
			mp3File.save(dir + "/" + artist + "/" + album + "/" + file + ".mp3");
		}
	}
}

File: sample.go
package main
import (
	// "fmt"
	"github.com/wtolson/go-taglib"
	"strings"
	"os"
)
func main() {
	var artistList, albumList, songList, genreList []string
	musicDir := os.Getenv("HOME") + "/" + "Music"
	genreList, err := listDir(musicDir)
	if err != nil {
		panic(err)
	}
	for _, absGenre := range(genreList) {
		// fmt.Println("genre:\t" + absGenre)

		artistList, err = listDir(musicDir + "/" + absGenre)
		if err != nil {
			panic(err)
		}
		for _, absArtist := range(artistList) {
			// fmt.Println("artist:\t" + absArtist)
			songList, err = listMusic(musicDir, absGenre, absArtist, "")
			if err != nil {
				panic(err)
			}
			editMusic(musicDir, absGenre, absArtist, "", songList)
			albumList, err = listDir(musicDir + "/" + absGenre + "/" + absArtist)
			if err != nil {
				panic(err)
			}
			for _, absAlbum := range(albumList) {
				// fmt.Println("album:\t-\t" + absAlbum)
				songList, err = listMusic(musicDir, absGenre, absArtist, absAlbum)
				if err != nil {
					panic(err)
				}
				editMusic(musicDir, absGenre, absArtist, absAlbum, songList)
			}
		// fmt.Println()
		}
	}
}

func listDir(dir string) (subDirs []string, err error) {
	fileDir, err := os.Open(dir)
	if err != nil {
		return
	}
	files, err := fileDir.Readdir(-1)
	if err != nil {
		return
	}
	for _, file := range files {
		if file.IsDir() {
			subDirs = append(subDirs, file.Name())
		}
	}
	return
}

func listMusic(musicDir, genre, artist, album string) (musicList []string, err error) {
	var currentDir *os.File
	if strings.Compare(album, "") == 0 {
		currentDir, err = os.Open(musicDir + "/" + genre + "/" + artist)
		if err != nil {
			return musicList, err
		}
	} else {
		currentDir, err = os.Open(musicDir + "/" + genre + "/" + artist + "/" + album)
		if err != nil {
			return musicList, err
		}
	}
	files, err := currentDir.Readdir(-1)
	if err != nil {
		return musicList, err
	}
	for _, file := range files {
		if len(file.Name()) > 3 {
			if strings.Compare(file.Name()[len(file.Name()) - 4:], ".mp3") == 0 {
				// fmt.Println(file.Name())
				musicList = append(musicList, file.Name())
			}
		}
	}
	return
}

func editMusic (musicDir, genre, artist, album string, songs []string) (err error) {
	var file *taglib.File
	for _, absSong := range(songs) {
		// fmt.Println("song:\t-\t-\t" + absSong)
		if album == "" {
			file, err = taglib.Read(musicDir + "/" + genre + "/" + artist + "/" + absSong)
		} else {
			file, err = taglib.Read(musicDir + "/" + genre + "/" + artist + "/" + album + "/" + absSong)
		}
		if err != nil {
			return err
		}
		if file.Genre() != genre {
			file.SetGenre(genre)
		}
		if file.Artist() != artist {
			file.SetArtist(artist)
		}
		if album == "" {
			file.SetAlbum("[non-album track]")
		} else {
			file.SetAlbum(album)
		}
		if file.Title() != absSong[:len(absSong) - 4] {
			file.SetTitle(absSong[:len(absSong) - 4])
		}
		err = file.Save()
		if err != nil {
			return err
		}
		file.Close()
	}
	return err
}
